tcljava/doc/Notifier.3
'\"
'\" Copyright (c) 1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id: Notifier.3,v 1.1 1998/10/14 21:09:09 cvsadmin Exp $
'\" 
.so man.macros
.TH Notifier 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.Notifier -- The event queue and notifier API.

.SH METHODS
.sp
synchronized void \fBdeleteEvents\fR(EventDeleter \fIdeleter\fR)
.sp
synchronized int \fBdoOneEvent\fR(int \fIflags\fR)
.sp
static synchronized Notifier \fBgetNotifierForThread\fR(Thread
\fIthread\fR)
.sp
synchronized void \fBpreserve\fR()
.sp
synchronized void \fBqueueEvent\fR(TclEvent \fIevent\fR, int \fIpos\fR)
.sp
synchronized void \fBrelease\fR()

.SH ARGUMENTS
.AS TclObject "table[]" .
.AP Thread thread
The thread in which the Notifier should run.
.AP EventDeleter deleter
The EventDeleter determines which event in the event queue should be
deleted.
.AP TclEvent event
The event to put into the event queue.
.AP int pos
The position at which the event should be queued.
.AP int flags
What types of events to service.
.BE

.SH DESCRIPTION

The \fBNotifier\fR is the lowest-level part of the Jacl event
system. It is used by higher-level event sources such as file,
JavaBean and timer events. The Notifier manages an event queue that
holds \fBTclEvent\fR objects.

The Jacl \fBNotifier\fR is designed to run in a multi-threaded
environment. Each \fBNotifier\fR instance is associated with a
\fBprimary thread\fR. Any thread can queue (or dequeue) events using
the \fBqueueEvent\fR (or \fBdeleteEvents\fR) call. However, only the
primary thread may process events in the queue using the
\fBdoOneEvent\fR call. Attempts to call \fBdoOneEvent\fR from a
non-primary thread will cause a \fBTclRuntimeError\fR.

This class doesn't have a public constructor and cannot be directly
instantiated. You can get to it only through
\fBNotifier.getNotifierForThread\fR or the \fBgetNotifier\fR method of
the \fBInterp\fR class.

.TP
\fBgetNotifierForThread\fR
This static method returns the \fBNotifier\fR instance whose primary
thread is the given thread. If such an instance does not yet exist, it
is created automatically.
.TP
\fBpreserve\fR, \fBrelease\fR
These two methods maintain a reference count to manage the lifetime of
a Notifier and its corresponding event queue. When a Notifier is
created inside the \fB\fBgetNotifierForThread\fR method, its reference
count is zero. A call to preserve increments the reference count by
one; a call to \fBrelease\fR decrements the reference count by
one. When the reference count goes from one to zero, the Notifier is
disposed -- its event queue is deleted and any left-over events in the
queue will not be processed.
.sp
If a Java module intends to use a Notifier returned by
\fBgetNotifierForThread\fR over one or more method calls, it should
call \fBpreserve\fR immediately after \fBgetNotifierForThread\fR
returns. The module should call \fBrelase\fR when it no longer needs
access to the notifier.
.sp
NOTE: the notifier returned by \fIinterp\fR.\fBgetNotifier\fR will remain
valid as long as the Interpreter is valid. Hence, if a module always
access a notifier via calls to \fIinterp\fR.\fBgetNotifier\fR, it needs not
use the notifier's \fBpreserve\fR and \fBrelease\fR methods.
.TP
\fBqueueEvent\fR
This method puts the \fIevent\fR into the event queue at the position
specified by the \fIpos\fR argument. The \fIpos\fR argument may be one
of the following:
.RS
.TP
\fBTCL.QUEUE_TAIL\fR
Add the event at the back of the queue, so that all other pending
events will be serviced first.  This is almost always the right place
for new events.
.TP
\fBTCL.QUEUE_HEAD\fR
Add the event at the front of the queue, so that it will be serviced
before all other queued events. 
.TP
\fBTCL.QUEUE_MARK\fR
Add the event at the front of the queue, unless there are other events
at the front whose position is \fBTCL.QUEUE_MARK\fR; if so, add the
new event just after all other \fBTCL.QUEUE_MARK\fR events.  This
value of position is used to insert an ordered sequence of events at
the front of the queue, such as a series of Enter and Leave events
synthesized during a grab or ungrab operation in Tk.
.RE
.TP
\fBdeleteEvents\fR
This method provides a way to selectively delete events in the event
queue. It calls \fIdeleter\fR.\fBdeleteEvent\fR for each event in the
queue and deletes those for which \fIdeleter\fR.\fBdeleteEvent\fR
returns 1. Events for which the \fIdeleter\fR returns 0 are left in the
queue.
.TP
\fBdoOneEvent\fR
This method is called in the context of the Notifier's primary thread
to dispatch events in the event queue. If events are found in the
event queue, the \fBprocessEvent\fR method of the first available
event is called. If no event are in the queue, \fBdoOneEvent\fR checks
if there are idle handlers. If there are any, it invokes all of them
and returns. Finally, if no events or idle callbacks have been found,
then \fBdoOneEvent\fR sleeps until an event occurs, calles
\fBprocessEvent\fR the of the event and returns.
.sp
If the \fIflags\fR argument to \fBdoOneEvent\fR is non-zero, it
restricts the kinds of events that will be processed by
\fBdoOneEvent\fR. Flags may be an OR-ed combination of any of the
following bits:
.RS
.TP
\fBTCL.WINDOW_EVENTS\fR
Process window system events.
.TP
\fBTCL.FILE_EVENTS\fR
Process file events.
.TP
\fBTCL.TIMER_EVENTS\fR
Process timer events.
.TP
\fBTCL.IDLE_EVENTS\fR
Process idle callbacks.
.TP
\fBTCL.ALL_EVENTS\fR
Process all kinds of events: equivalent to OR-ing together all of the
above flags or specifying none of them.
.TP
\fBTCL.DONT_WAIT\fR
Don't sleep: process only events that are ready at the time of the
call.
.PP
If any of the flags \fBTCL.WINDOW_EVENTS\fR, \fBTCL.FILE_EVENTS\fR,
\fBTCL.TIMER_EVENTS\fR, or \fBTCL.IDLE_EVENTS\fR is set, then the only
events that will be considered are those for which flags are set.
Setting none of these flags is equivalent to the value
\fBTCL.ALL_EVENTS\fR, which causes all event types to be processed.
If an application has additional event sources then additional flag
values may also be valid, depending on those event sources.
.sp
The \fBTCL.DONT_WAIT\fR flag causes \fBdoOneEvent\fR not to put the
process to sleep: it will check for events but if none are found then
it returns immediately with a return value of 0 to indicate that no
work was done. \fBdoOneEvent\fR will also return 0 without doing
anything if the only alternative is to block forever (this can happen,
for example, if flags is \fBTCL.IDLE_EVENTS\fR and there are no idle
callbacks pending, or if no event handlers or timer handlers exist).
.sp
\fBdoOneEvent\fR may be invoked recursively. For example, it is
possible to invoke \fBdoOneEvent\fR recursively from an event handler
called by \fBdoOneEvent\fR. This sort of operation is useful in some
modal situations, such as when a notification dialog has been popped
up and an application wishes to wait for the user to click a button in
the dialog before doing anything else.
.RE

.SH "EQUIVALENT C FUNCTIONS"
Tcl_QueueEvent, Tcl_DeleteEvents, Tcl_DoOneEvent

.SH "SEE ALSO"
TclEvent, TimerHandler, IdleHandler, EventDeleter, Interp

.SH KEYWORDS
event, notifier, event queue, event sources, file events, timer, idle


tcljava/ChangeLog
2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * README.jacl:
        * README.tclblend: Mention that Jacl and Tcl Blend
        are production ready and include URLs for build
        notes on the wiki.

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        Update version numbers to 1.3.3.

        * configure:
        * configure.in:
        * src/pkgIndex.tcl:
        * src/jacl/tcl/lang/Interp.java:
        * src/tcljava/tcl/lang/BlendExtension.java:
        * unix/tools/vutil/current:
        * win/makefile.vc:

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * configure: Regen.
        * configure.in: Set default --prefix to /usr/local.

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/tclblend/tcl/lang/Interp.java (evalFile):
        Add TclBlend implementation of Interp.evalFile().

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/jacl/tcl/lang/ExecCmd.java (cmdProc):
        Reimplement JDK 1.3 and newer exec so that two
        threads are created. One reads stdout of the
        subprocess while the other reads stderr. This
        is needed to avoid deadlocking in waitFor()
        when a process writes a lot of stdout and stderr.
        * src/tests/tcljava/tests/exec/ExecLotsOutput.java:
        * tests/inprogress/ExecCmd.test: Add test case
        for the waitFor() exec deadlocking bug.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-branch branch and
 	and tcljava-132-branchpoint tags.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-2005-08-08 tag to mark tree
	state for the 1.3.2 release.

2005-07-28  Mo DeJong  <mdejong@users.sourceforge.net>

	* announce: Update.
	* docs/website/index.html: Update.

2005-07-27  Mo DeJong  <mdejong@users.sourceforge.net>

	* README.jacl: Update msys description and
	mention jaclsh.bat script.
	* README.tclblend: Mention Tcl 8.4.11 and
	Thread 2.6.2 reqs. Add msys url and
	mention jtclsh.bat script.
	* docs/website/faq.html: Update FAQ.
	* docs/website/index.html: Update homepage
	info with new 1.3.2 release.
	* docs/website/mail.html: Add mail page that
	explains how the mailing lists work.

2005-07-23  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Channel.java (initOutput):
	When initializing a channel of type "file",
	set a flag to indicate that the channel
	should be sync()'ed in addition to flushing.
	* src/jacl/tcl/lang/FlushCmd.java: Print
	IO exception stack trace, this should never
	happen but if it did sa stack trace would
	be helpful.
	* src/jacl/tcl/lang/TclOutputStream.java (setSync,
	flushChannel):
	Add flag setting method so that an output
	stream can be marked so that sync() will
	be invoked when flushing.
	* tests/tcl/SocketCmd.test: Add simple socket
	test case that writes to and reads from a socket.

2005-07-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/BinaryCmd.java (FormatNumber,
	ScanNumber):
	Handle edge cases when casting from a double
	to a float. Rewrite scanning code so that
	64 bit long value is used for shifting operation.
	* src/jacl/tcl/lang/InfoCmd.java:
	* src/jacl/tcl/lang/Shell.java:
	* src/jacl/tcl/lang/SubstCmd.java:
	Invoke static method via class name.
	* src/jacl/tcl/lang/Util.java (strtod): Add
	support for "NaN", "Inf", and "-Inf" strings
 	as valid double values.
	* tests/tcl/binary.test: Enable test cases
	that were not being run because they had
	the nonPortable constraint. Java supports
	only big endian primitive types so the
	binary format and scan operations can
	just assume big endian layout. Test that
	assume little endian are not run via
	addition of the littleEndian constraint.
	Add test cases for double -> float edge cases.

2005-07-20  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Copy new files into dist.
	* tests/tclblend/loading.test: Rework test
	so that it is only run under Win32 and
	returns the proper result.
	* win/jaclsh.bat: Removed old bat script.

2005-07-19  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Remove old green threads code
	for "custom" target. Tcl Blend only works
	with a native threads JVM anyway. Add new
	custom2 shell which loads a JVM at runtime
	and then loads Tcl Blend like a JVM would.
	This shell make it easier to debug weird
	loading issues in Tcl Blend.
	* src/tests/tclblend/tests/JavaEval.java:
	Make the program exit(0) when an expected
	result matches the program result. Use
	positive exit status.
	* tests/tclblend/loading.test: Use the
	expected compare test in the JavaEval
	class.
	* unix/custom2.c: Add new custom shell impl
	that loads a JVM and then loads Tcl Blend into
	the JVM.

2005-07-18  Mo DeJong  <mdejong@users.sourceforge.net>

	Fix crashes when loading Tcl Blend into an
	existing JVM by invoking Tcl_FindExecutable("java")
	to init Tcl subsystems.

	* src/native/javaCmd.c: Update comment.
	* src/native/javaInterp.c (
	Java_tcl_lang_Interp_create,
	Java_tcl_lang_Interp_init,
	Java_tcl_lang_Interp_initName): Add
	TCLBLEND_DEBUG output statements.
	Add initName() method that will init the
	Tcl executable name when Tcl Blend is loaded
	from Java.
	* src/tclblend/tcl/lang/Interp.java (shlib_load,
	initName):
	Add initName() method mentioned above and
	invoke it after loading the Tcl Blend shared lib.
	* tests/tclblend/loading.test: Update tests
	to account for executable name check. Add
	test for [info nameofexecutable] in a Tcl Blend
	interp loaded into Java.

2005-07-17  Mo DeJong  <mdejong@users.sourceforge.net>

	Add startup .bat files for Jacl and Tcl Blend.
	The bat files are auto generated by configure
	with the correct paths.

	* Makefile.in: Install and test jaclsh.bat
	and jtclsh.bat. Make sure the PATH variable
	is always set with the new directories
	prepended to the PATH. Rework tricky gdb
	invocation under Win32 since the mingw
	version of gdb does not know about mingw
	style unix paths.

tcljava/src/tclblend/tcl/lang/Notifier.java
/*
 * Notifier.java --
 *
 *	Implements the Blend version of the Notifier class.
 *
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: Notifier.java,v 1.3 2000/10/29 06:00:42 mdejong Exp $
 *
 */

package tcl.lang;

import java.util.Hashtable;
import java.util.Vector;

// Implements the Tcl Blend version of the Notifier class. The Notifier is
// the lowest-level part of the event system. It is used by
// higher-level event sources such as file, JavaBean and timer
// events. The Notifier manages an event queue that holds TclEvent
// objects.
//
// The Tcl Blend Notifier is designed to run in a multi-threaded
// environment. Each notifier instance is associated with a primary
// thread. Any thread can queue (or dequeue) events using the
// queueEvent (or deleteEvents) call. However, only the primary thread
// may process events in the queue using the doOneEvent()
// call. Attepmts to call doOneEvent() from a non-primary thread will
// cause a TclRuntimeError.
//
// This class does not have a public constructor and thus cannot be
// instantiated. The only way to for a Tcl extension to get an
// Notifier is to call Interp.getNotifier() (or
// Notifier.getNotifierForThread() ), which returns the Notifier for that
// interpreter (thread).

public class Notifier implements EventDeleter {

private static Notifier globalNotifier;

// First pending event, or null if none.

private TclEvent firstEvent;

// Last pending event, or null if none.

private TclEvent lastEvent;

// Last high-priority event in queue, or null if none.

private TclEvent markerEvent;

// The primary thread of this notifier. Only this thread should process
// events from the event queue.

Thread primaryThread;

// Reference count of the notifier. It's used to tell when a notifier
// is no longer needed.

int refCount;

// Stores the Notifier for each thread.

private static Hashtable notifierTable = new Hashtable();

// tcl ThreadId

private long tclThreadId;

// Mutex used to protect concurrent access to the internals of this Notifier
// object.  For example, the queueing and dequeueing of objects from the
// event list.

private final Object notifierMutex = new Object();


/*
 *----------------------------------------------------------------------
 *
 * Notifier --
 *
 *	Creates a Notifier instance.
 *
 * Side effects:
 *	Member fields are initialized.
 *
 *----------------------------------------------------------------------
 */

private
Notifier(
    Thread primaryTh)		// The primary thread for this Notifier.
{
    primaryThread     = primaryTh;
    firstEvent        = null;
    lastEvent         = null;
    markerEvent       = null;
    tclThreadId       = init();
    refCount          = 0;
}

/*
 *----------------------------------------------------------------------
 *
 * getNotifierForThread --
 *
 *	Get the notifier for this thread, creating the Notifier,
 *	when necessary.
 *
 * Results:
 *	The Notifier for this thread.
 *
 * Side effects:
 *	The Notifier is created when necessary.
 *
 *----------------------------------------------------------------------
 */

public static synchronized Notifier
getNotifierForThread(
    Thread thread)		// The thread that owns this Notifier.
{
    Notifier notifier = (Notifier) notifierTable.get(thread);
    if (notifier == null) {
	notifier = new Notifier(thread);
	notifierTable.put(thread, notifier);
    }

    return notifier;
}

/*
 *----------------------------------------------------------------------
 *
 * preserve --
 *
 *	Increment the reference count of the notifier. The notifier will
 *	be kept in the notifierTable (and alive) as long as its reference
 *	count is greater than zero.  This method is concurrent safe.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The refCount is incremented.
 *
 *----------------------------------------------------------------------
 */

public void
preserve()
{
    synchronized (notifierMutex) {
        if (refCount < 0) {
	    throw new TclRuntimeError(
	        "Attempting to preserve a freed Notifier");
        }
        ++refCount;
    }
}

/*
 *----------------------------------------------------------------------
 *
 * release --
 *
 *	Decrement the reference count of the notifier. The notifier will
 *	be freed when its refCount goes from one to zero.  This method is
 *	concurrent safe.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The notifier may be removed from the notifierTable when its
 *	refCount reaches zero.
 *
 *----------------------------------------------------------------------
 */

public void
release()
{
    synchronized (notifierMutex) {
// FIXME: indent this block properly later.
    if ((refCount == 0) && (primaryThread != null)) {
	throw new TclRuntimeError(
		"Attempting to release a Notifier before it's preserved");
    }
    if (refCount <= 0) {
	throw new TclRuntimeError("Attempting to release a freed Notifier");
    }
    --refCount;
    if (refCount == 0) {
	notifierTable.remove(primaryThread);
	primaryThread = null;
	dispose();
    }
    }
}

/*
 *----------------------------------------------------------------------
 *
 * queueEvent --
 * 
 *	Insert an event into the event queue at one of three
 *	positions: the head, the tail, or before a floating marker.
 *	Events inserted before the marker will be processed in
 *	first-in-first-out order, but before any events inserted at
 *	the tail of the queue.  Events inserted at the head of the
 *	queue will be processed in last-in-first-out order.  This method is
 *	concurrent safe. 
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	If this method is invoked by a non-primary thread, the
 *	primaryThread of this Notifier will be notified about the new
 *	event.
 *
 *----------------------------------------------------------------------
 */

public void
queueEvent(
    TclEvent evt,		// The event to put in the queue.
    int position)		// One of TCL.QUEUE_TAIL,
				// TCL.QUEUE_HEAD or TCL.QUEUE_MARK.
{
    evt.notifier = this;

    synchronized (notifierMutex) {
//FIXME: indent this block properly
    if (position == TCL.QUEUE_TAIL) {
	// Append the event on the end of the queue.
	evt.next = null;

	if (firstEvent == null) {
	    firstEvent = evt;
	} else {
	    lastEvent.next = evt;
	}
	lastEvent = evt;
    } else if (position == TCL.QUEUE_HEAD) {
	// Push the event on the head of the queue.
	evt.next = firstEvent;
	if (firstEvent == null) {
	    lastEvent = evt;
	}
	firstEvent = evt;
    } else if (position == TCL.QUEUE_MARK) {
	// Insert the event after the current marker event and advance
	// the marker to the new event.
	if (markerEvent == null) {
	    evt.next = firstEvent;
	    firstEvent = evt;
	} else {
	    evt.next = markerEvent.next;
	    markerEvent.next = evt;
	}
	markerEvent = evt;
	if (evt.next == null) {
	    lastEvent = evt;
	}
    } else {
	// Wrong flag.
	throw new TclRuntimeError("wrong position \"" + position +
	       "\", must be TCL.QUEUE_HEAD, TCL.QUEUE_TAIL or TCL.QUEUE_MARK");
    }
    }

    if (Thread.currentThread() != primaryThread) {
	alertNotifier(tclThreadId);
    }
}

/*
 *----------------------------------------------------------------------
 *
 * deleteEvents --
 *
 *	Calls an EventDeleter for each event in the queue and deletes
 *	those for which deleter.deleteEvent() returns 1. Events
 *	for which the deleter returns 0 are left in the queue.   This 
 *	method is concurrent safe. 
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Potentially removes one or more events from the event queue.
 *
 *----------------------------------------------------------------------
 */

public void
deleteEvents(
    EventDeleter deleter)	// The deleter that checks whether an event
				// should be removed.
{
    TclEvent evt, prev;

    synchronized (notifierMutex) {
//FIXME: indent this block properly
    for (prev = null, evt = firstEvent; evt != null; evt = evt.next) {
        if (deleter.deleteEvent(evt) == 1) {
            if (firstEvent == evt) {
                firstEvent = evt.next;
                if (evt.next == null) {
                    lastEvent = null;
                }
		    if (markerEvent == evt) {
			markerEvent = null;
		    }
            } else {
                prev.next = evt.next;
		    if (evt.next == null) {
			lastEvent = prev;
            }
	    if (markerEvent == evt) {
			markerEvent = prev;
		    }
	    }
        } else {
            prev = evt;
        }
    }
    }
}

/*
 *----------------------------------------------------------------------
 *
 * deleteEvent --
 *
 *	This method is required for this class being an EventDeleter.
 *	Checks the given event to see if it is already processed.
 *	This method together with 'deleteEvents' are used by serviceEvent to
 *	remove an event that is just processed in a concurrent safe manor.
 *
 * Results:
 *	Returns 1 if the given event is processed, 0 otherwise.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

public int
deleteEvent(
    TclEvent evt) {		// Check whether this event should be removed.
    return ((evt.isProcessing == true) &&
	    (evt.isProcessed  == true)) ? 1 : 0;
}

/*
 *----------------------------------------------------------------------
 *
 * serviceEvent --
 *
 *	Process one event from the event queue.  This method is concurrent 
 *	safe and can be reentered recursively.
 *
 * Results:
 *	The return value is 1 if the procedure actually found an event
 *	to process. If no processing occurred, then 0 is returned.
 *
 * Side effects:
 *	Invokes all of the event handlers for the highest priority
 *	event in the event queue.  May collapse some events into a
 *	single event or discard stale events.
 *
 *----------------------------------------------------------------------
 */
int
serviceEvent(
    int flags)			// Indicates what events should be processed.
				// May be any combination of TCL.WINDOW_EVENTS
				// TCL.FILE_EVENTS, TCL.TIMER_EVENTS, or other
				// flags defined elsewhere.  Events not
				// matching this will be skipped for processing
				// later.
{
    TclEvent evt, prev;

    // No event flags is equivalent to TCL_ALL_EVENTS.
    
    if ((flags & TCL.ALL_EVENTS) == 0) {
	flags |= TCL.ALL_EVENTS;
    }

    // Loop through all the events in the queue until we find one
    // that can actually be handled.

    evt = null;
    while ((evt = getAvailableEvent(evt)) != null) {
	// Call the handler for the event.  If it actually handles the
	// event then free the storage for the event.  There are two
	// tricky things here, both stemming from the fact that the event
	// code may be re-entered while servicing the event:
	//
	// 1. Set the "isProcessing" field to true. This is a signal to
	//    ourselves that we shouldn't reexecute the handler if the
	//    event loop is re-entered.
	// 2. When freeing the event, must search the queue again from the
	//    front to find it.  This is because the event queue could
	//    change almost arbitrarily while handling the event, so we
	//    can't depend on pointers found now still being valid when
	//    the handler returns.
	synchronized (evt) {
//FIXME: indent this properly
	boolean b = evt.isProcessing;
	evt.isProcessing = true;

	if ((b == false) && (evt.processEvent(flags) != 0)) {
	    evt.isProcessed = true;
	    if (evt.needsNotify) {
		    evt.notifyAll();
		}
		
		deleteEvents(this);	// this takes care of deletion of the
					// just processed event
	    return 1;
	} else {
	    // The event wasn't actually handled, so we have to
	    // restore the isProcessing field to allow the event to be
	    // attempted again.

	    evt.isProcessing = b;

	// The handler for this event asked to defer it.  Just go on to
		// the next event.  we will try to find another event to
		// process when the while loop continues
	    }
	}
    }

    return 0;
}

/*
 *----------------------------------------------------------------------
 *
 * getAvailableEvent --
 *
 *	Search through the internal event list to find the first event
 *	that is has not being processed AND the event is not equal to the given
 *	'skipEvent'.  This method is concurrent safe.
 *	
 * Results:
 *	The return value is a pointer to the first found event that can be
 * 	processed.  If no event is found, this method returns null.
 *
 * Side effects:
 *	This method synchronizes on the 'notifierMutex', which will block any
 *	other thread from adding or removing events from the event queue.
 *
 *----------------------------------------------------------------------
 */
    
private TclEvent
getAvailableEvent(
    TclEvent    skipEvent)	// Indicates that the given event should not
				// be returned.  This argument can be null.
{
    TclEvent evt;
    
    synchronized(notifierMutex) {
	for (evt = firstEvent ; evt != null ; evt = evt.next) {
	    if ((evt.isProcessing == false) &&
		(evt.isProcessed  == false) &&
		(evt != skipEvent)) {
		return evt;
	    }
	}
    }
    return null;
}

/*
 *----------------------------------------------------------------------
 *
 * doOneEvent --
 *
 *	Process a single event of some sort.  If there's no work to
 *	do, wait for an event to occur, then process it. May delay
 *	execution of process while waiting for an event, unless
 *	TCL.DONT_WAIT is set in the flags argument.  This routine
 *	should only be called from the primary thread for the
 *	notifier.
 *	
 * Results:
 *	The return value is 1 if the procedure actually found an event
 *	to process. If no processing occurred, then 0 is returned
 *	(this can happen if the TCL.DONT_WAIT flag is set or if there
 *	are no event handlers to wait for in the set specified by
 *	flags).
 *
 * Side effects:
 *	May delay execution of process while waiting for an event,
 *	unless TCL.DONT_WAIT is set in the flags argument. Event
 *	sources are invoked to check for and queue events. Event
 *	handlers may produce arbitrary side effects.
 *
 *----------------------------------------------------------------------
 */

public native int
doOneEvent(
    int flags);			// Miscellaneous flag values: may be any
				// combination of TCL.DONT_WAIT,
				// TCL.WINDOW_EVENTS, TCL.FILE_EVENTS,
				// TCL.TIMER_EVENTS, TCL.IDLE_EVENTS,
				// or others defined by event sources.

/*
 *----------------------------------------------------------------------
 *
 * alertNotifier --
 *
 *	Wake up the C notifier.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	May cause the C level event loop to wake up.
 *
 *----------------------------------------------------------------------
 */

private final native void
alertNotifier(long tid);


/*
 *----------------------------------------------------------------------
 *
 * init --
 *
 *	Initialize the C event source for the Java notifier.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

private final native long
init();

/*
 *----------------------------------------------------------------------
 *
 * dispose --
 *
 *	Tear down the C event source for the Java notifier.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

final native void
dispose();

/*
 *----------------------------------------------------------------------
 *
 * hasEvents --
 *
 *	Check to see if there are events waiting to be processed.  This
 *	method is concurrent safe.
 *
 * Results:
 *	Returns true if there are events on the Notifier queue.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

private final boolean 
hasEvents()
{
    boolean result = (getAvailableEvent(null) != null);

    return result;
}

} // end Notifier
